extensions/PhabBugz/lib/Util.pm

```diff
@@ -0,0 +1,211 @@
+    my $project = shift;
+
+    my $data = {
+        queryKey => 'active',
+        constraints => {
+            name => $project
+        }
+    };
+
+    my $result = request('project.search', $data);
+    if (!$result->{result}{data}) {
+        return undef;
+    }
+    return $result->{result}{data}[0]{phid};
+}
+
+sub create_project {
+    my ($project, $description, $members) = @_;
+
+    my $data = {
+        transactions => [
+            { type => 'name',  value => $project           },
+            { type => 'description', value => $description },
+            { type => 'edit',  value => 'admin'            },
+            { type => 'join',  value => 'admin'            },
+            { type => 'view',  value => 'admin'            },
+            { type => 'icon',  value => 'group'            },
+            { type => 'color', value => 'red'              }
+        ]
+    };
+
+    my $result = request('project.edit', $data);
+    return $result->{result}{object}{phid};
+}
+
+sub set_project_members {
+    my ($project_id, $phab_user_ids) = @_;
+
+    my $data = {
+        objectIdentifier => $project_id,
+        transactions => [
+            { type => 'members.set',  value => $phab_user_ids }
+        ]
+    };
+
+    my $result = request('project.edit', $data);
+    return $result->{result}{object}{phid};
+}
+
+sub get_members_by_bmo_id {
+    my $users = shift;
+
+    my $data = {
+        accountids => [ map { $_->id } @$users ]
+    };
+
+    my $result = request('bmoexternalaccount.search', $data);
+    if (!$result->{result}) {
+        return [];
+    }
+
+    my @phab_ids;
+    foreach my $user (@{ $result->{result} }) {
+        push(@phab_ids, $user->{phid});
+    }
+    return \@phab_ids;
+}
+
+sub request {
+    my ($method, $data) = @_;
+
+    my $phab_api_key  = Bugzilla->params->{phabricator_api_key};
+    my $phab_base_uri = Bugzilla->params->{phabricator_base_uri};
+    $phab_base_uri || ThrowUserError('invalid_phabricator_uri');
+    $phab_api_key  || ThrowUserError('invalid_phabricator_api_key');
+
+    state $ua;
+
+    if (!$ua) {
+        $ua = LWP::UserAgent->new(timeout => 10);
+        if (Bugzilla->params->{proxy_url}) {
+            $ua->proxy('https', Bugzilla->params->{proxy_url});
+        }
+        $ua->default_header('Content-Type' => 'application/x-www-form-urlencoded');
+    }
+
+    my $full_uri = $phab_base_uri . '/api/' . $method;
+
+    $data->{__conduit__} = { token => $phab_api_key };
+
+    my $response = eval {
+        $ua->post($full_uri, { params => encode_json($data) });
+    };
+
+    $response->is_error
+        && ThrowCodeError('phabricator_api_error',
+                          { reason => $response->message });
+
+    my $result = decode_json($response->content);
+    if ($result->{error_code}) {
+        ThrowCodeError('phabricator_api_error',
+                       { code   => $result->{error_code},
+                         reason => $result->{error_info} });
+    }
```

Comment:
This is border-line -- the arguments to ThrowCodeError make it unsightly to use postfix if, so this can stay.